src/pcrlock/pcrlock.c

```diff
@@ -4567,15 +4567,58 @@ static int make_policy(bool force, bool recovery_pin) {
                 log_info("Retrieved PIN from TPM2 in %s.", FORMAT_TIMESPAN(usec_sub_unsigned(now(CLOCK_MONOTONIC), pin_start_usec), 1));
         }
 
-        TPM2B_NV_PUBLIC nv_public = {};
+        /* Load the PIN (which we have stored in the "auth" TPM2B_AUTH) into the TPM as an HMAC key so that
+         * we can use it in a TPM2_PolicySigned() to write to the nvindex */
+        _cleanup_(tpm2_handle_freep) Tpm2Handle *pin_handle = NULL;
+
+        TPM2B_PUBLIC auth_hmac_public = {
+                .size = sizeof(TPMT_PUBLIC),
+                .publicArea = {
+                        .type = TPM2_ALG_KEYEDHASH,
+                        .nameAlg = TPM2_ALG_SHA256,
+                        .objectAttributes = TPMA_OBJECT_DECRYPT | TPMA_OBJECT_SIGN_ENCRYPT /* | TPMA_OBJECT_USERWITHAUTH */,
+                        .parameters.keyedHashDetail.scheme.scheme = TPM2_ALG_SHA256,
+                        .unique.keyedHash.size = auth.size,
```

Comment:
Unless I'm missing something, I don't see anything that calculates this field.

Reply:
I think this was pointed out, but keyedhash detail scheme scheme should be hmac and scheme.halg or something like that should be hash, ie sha256.

Reply:
Took me awhile, but the hmac private seems to have a secret and a seed. Those are used to calculate the unique field. The python code does it here:
  - https://github.com/tpm2-software/tpm2-pytss/blob/19aee3fbfd5acfcdb59ff7c920fc75ac89817c8f/src/tpm2_pytss/types.py#L2148

Reply:
Yeah, see the comment here as well https://github.com/systemd/systemd/pull/31790#issuecomment-2059134174.

I was wondering whether the seed field might be optional in this context because asymmetric keys that aren't storage objects don't have one but it looks like `CryptValidateKeys` in the reference implementation requires that the seed equals the size of the name algorithm for symmetric keys.

Reply:
Yeah it needs it, but it can be whatever you want. IIRC it's used to bind symmetric objects  Public and Private portions via unique.